apps/src/blockly/addons/contextMenu.ts

```diff
@@ -313,22 +362,47 @@ export function registerHelp() {
     },
     scopeType: GoogleBlockly.ContextMenuRegistry.ScopeType.BLOCK,
     id: 'blockHelp',
-    weight: 11,
+    weight,
   };
   GoogleBlockly.ContextMenuRegistry.registry.register(helpOption);
 }
 
+function registerAllCursors(startingWeight: number, cursorTypes: any[]) {
+  cursorTypes.forEach((cursorType, index) => {
+    const weight = startingWeight + index;
+    registerCursor(cursorType, weight);
+  });
+}
+
+function isCurrentCursor(cursorType: string) {
+  const currentCursorType = Blockly.navigationController.cursorType;
+  return cursorType === currentCursorType;
+}
+
+function setNewCursor(type: string, scope: ContextMenuRegistry.Scope) {
+  localStorage.setItem(BLOCKLY_CURSOR, type);
+  Blockly.navigationController.cursorType = type;
+  const markerManager = Blockly.getMainWorkspace().getMarkerManager();
+  const oldCurNode = markerManager.getCursor()?.getCurNode();
+  Blockly.getMainWorkspace()
+    .getMarkerManager()
+    .setCursor(Blockly.getNewCursor(type));
+  if (Blockly.getFunctionEditorWorkspace()) {
+    Blockly.getFunctionEditorWorkspace()
+      ?.getMarkerManager()
+      .setCursor(Blockly.getNewCursor(type));
+  }
+  if (oldCurNode) {
+    markerManager.getCursor()?.setCurNode(oldCurNode);
+  }
+  // Set the navigation state to the workspace and moves the cursor to the top block.
```

Comment:
for my information: if this moves the cursor to the top block, what does `setCurNode` do?

Reply:
As far as I understand it, `setCurNode` just sets the location of the marker - the actual blinking SVG.

Reply:
Are the marker and the cursor different? What I'm trying to understand is the comment says we move the cursor to the top block, but the marker is wherever it was before?

Reply:
Oh good question. I just realized we're doing two contradictory things. When we say `setCurNode(oldCurNode)`, we're basically saying that we should keep the cursor on the same node (e.g. stack/block/field/connection) when the user changes from one cursor type to another. However, we're also moving the cursor to the top block after that. I think the latter makes more sense for the purposes of the testing that Sam wants to conduct.

The code for the preservation of the current node comes from the [test playground](https://google.github.io/blockly-samples/plugins/keyboard-navigation/test/index.html) for the plugin: https://github.com/google/blockly-samples/blob/0d09c5d2dfd31d370ada8393cb002a39ef3d028e/plugins/keyboard-navigation/test/index.js#L72-L74

Will remove logic that cared about the old node!